### PR TITLE
Multiprocessing module for debugging

### PR DESCRIPTION
**Pull request by shanaxel42**
This PR:
-Creates a multiprocessing.pool wrapper class called Spool (starfish+pool, but definitely open for suggestions) that, when given n_processes=1 overrides the map function to use a non-mulitprocessed map. 
-Adds n_processes as a parameter in detect.py and pixel_spot_detector.py
-refactors imagestack.transform to use the spool class in a n_processes=1 scenario
-Modified a couple filter/detect tests to verify everything still works with spool. 

With these changes we should be able to attach the debugger to any Filter/detect step when n_processes=1.
 

fixes: #941

### CODE REVIEW

File: starfish/imagestack/imagestack.py [hunk 1]
```diff
@@ -847,30 +847,24 @@ def transform(
...
-                               self._data._data.dtype),)) as pool:
-                mp_results = pool.imap(mp_applyfunc, selectors_and_slice_lists)
-                return list(zip(mp_results, selectors))
+        with Spool(
+                processes=n_processes,
+                initializer=SharedMemory.initializer,
+                initargs=((self._data._backing_mp_array,
+                           self._data._data.shape,
+                           self._data._data.dtype),)) as pool:
+            if n_processes == 1:
```
[the remark this answers is not in the document]
Comment (joshmoore) [reply starfish/imagestack/imagestack.py]: I haven't grok'd the underlying issue yet, but having run across [toolz](https://toolz.readthedocs.io/en/latest/parallelism.html following a [Rocklin blog post link](http://matthewrocklin.com/blog/work/2015/02/17/Towards-OOC-Bag), would an abstraction layer higher (one that potentially others are trying out with dask) help to determine when to use `multiprocessing`'s map versus regular `map` (versus perhaps `ipyparallel`'s map in the future)? From the config POV, it'd also be good to constrain the `num_processes==1` check to as few locations as possible so we can add an override, though if necessary that could be done on _reading_ `num_processes`.

[the remark this answers is not in the document]
Comment (ttung) [reply starfish/imagestack/imagestack.py]: `_multiprocessing_workflow` depends on `SharedMemory.initializer` being called.  In the `n_processes` > 1 case, that's handled by `mulitprocessing.Pool`.

One possible strategy is to have `Spool` mimic the logic of `multiprocessing.Pool()`s `initalizer` method when `n_processes` == 1.

[the remark this answers is not in the document]
Comment (shanaxel42) [reply starfish/imagestack/imagestack.py]: > One possible strategy is to have `Spool` mimic the logic of `multiprocessing.Pool()`s `initalizer` method when `n_processes` == 1.

This works!

File: starfish/imagestack/imagestack.py [hunk 2]
```diff
@@ -50,6 +49,7 @@
 from starfish.imagestack.parser.numpy import NumpyData
 from starfish.imagestack.parser.tileset import parse_tileset
 from starfish.intensity_table.intensity_table import IntensityTable
+from starfish.multiprocessing.pool import Spool
```
[the remark this answers is not in the document]
Comment (shanaxel42) [reply starfish/imagestack/imagestack.py]:
```starfish/multiprocessing/pool.py:5: error: Name 'Pool' already defined (possibly by an import)
starfish/multiprocessing/pool.py:5: error: Cycle in inheritance hierarchy```

We'd be breaking some inheritance logic, but we could change to the delegation strategy Tony mentioned if you wanted Pool

[the remark this answers is not in the document]
Comment (ambrosejcarr) [reply starfish/imagestack/imagestack.py]: Interesting. Thanks for exploring this. I suggest digging into why @ttung suggested that to start. It might be that the right pattern is actually governed by how we import these things, and we should name the class or function properly. The thing I _really_ care about is that the API is the same as the `stdlib` `multiprocessing.pool.Pool`. 

```python
# file: starfish/multiprocessing/pool.py

class OptionalPool:
    """Multiprocessing pool that executes on the main process when n_processes=1"""
    # code ...
```

then

```python
# file: starfish/image/filter/some_filter.py

from starfish.multiprocessing.pool import OptionalPool as Pool
# code ...
```

File: starfish/imagestack/imagestack.py [hunk 3]
```diff
@@ -847,44 +847,19 @@ def transform(
...
-            worker_callable: Callable[[np.ndarray], Any],
-            numpy_array: np.ndarray,
-            selector_and_slice_list: Tuple[Mapping[Axes, int],
-                                           Tuple[Union[int, slice], ...]],
-    ):
-        sliced = numpy_array[selector_and_slice_list[1]]
-
-        return worker_callable(sliced)
+                self._processing_workflow, partial(func, **kwargs))
+            results = pool.imap(mp_applyfunc, selectors_and_slice_lists)
```
Comment (ambrosejcarr) [inline starfish/imagestack/imagestack.py]: I notice we're using imap here, but we only wrap map in the multiprocessing module. Is this a problem?

File: starfish/multiprocessing/pool.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        mp.Pool.__init__(self, **kwargs)
+        self.n_processes = processes
+
+    def map(self, func, iterable, chunksize=None):
+        if self.n_processes == 1:
+            self._initializer(*self._initargs)
+            return map(func, iterable)
+        return mp.Pool.map(self, func, iterable, chunksize)
+
+    def __reduce__(self):
```
Comment (ambrosejcarr) [inline starfish/multiprocessing/pool.py]: Why is this needed?

File: starfish/imagestack/imagestack.py (hunk 3 above)
Comment (shanaxel42) [reply starfish/imagestack/imagestack.py]: ah good catch! It didn't matter with the two processing methods but now it does.

File: starfish/multiprocessing/pool.py (hunk 4 above)
Comment (shanaxel42) [reply starfish/multiprocessing/pool.py]: it's one of the abstract methods required for inheriting Pool

Comment (ttung) [reply starfish/multiprocessing/pool.py]: I'm confused.  It looks to be [implemented to prevent pool objects from being pickled](https://github.com/python/cpython/blob/master/Lib/multiprocessing/pool.py#L560).  When are we pickling?

File: starfish/multiprocessing/pool.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import multiprocessing.pool as mp
+from typing import Optional
+
+
+class Pool(mp.Pool):
+    """Wrapper class for multiprocessing pool. If n_processes=1 just use map on main
+    thread for debugging purposes """
+
+    def __init__(self, processes: Optional[int]=None, **kwargs):
```
Comment (ttung) [inline starfish/multiprocessing/pool.py]: To make this work exactly as multiprocessing.Pool, you would probably need:

`def __init__(self, processes: Optional[int]=None, *args, **kwargs):`

and then pass *args into mp.Pool's constructor.

File: starfish/multiprocessing/pool.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import multiprocessing.pool as mp
+from typing import Optional
+
+
+class Pool(mp.Pool):
+    """Wrapper class for multiprocessing pool. If n_processes=1 just use map on main
+    thread for debugging purposes """
```
Comment (ttung) [inline starfish/multiprocessing/pool.py]: main process.

it may also be worth documenting a couple of non-trivial differences in behavior between mp.Pool and this for n_processes==1:

1. Depending on the forking model, global state may not exist in mp.Pool whereas they do exist if we just map on main process..
2. initializer gets called on every invocation of mp.Pool.{imap,map}.
